searx/engines/__init__.py

```diff
@@ -1,169 +1,220 @@
+            and iso_lang != engine_lang
+            and not engine_lang.startswith(iso_lang)
+            and iso_lang not in engine.supported_languages
+        ):
+            engine.language_aliases[iso_lang] = engine_lang
 
     # language_support
-    setattr(engine, 'language_support', len(getattr(engine, 'supported_languages', [])) > 0)
+    engine.language_support = len(engine.supported_languages) > 0
 
     # assign language fetching method if auxiliary method exists
     if hasattr(engine, '_fetch_supported_languages'):
         headers = {
             'User-Agent': gen_useragent(),
             'Accept-Language': 'ja-JP,ja;q=0.8,en-US;q=0.5,en;q=0.3',  # bing needs a non-English language
         }
-        setattr(engine, 'fetch_supported_languages',
-                lambda: engine._fetch_supported_languages(get(engine.supported_languages_url, headers=headers)))
-
-    # tor related settings
-    if settings['outgoing'].get('using_tor_proxy'):
-        # use onion url if using tor.
-        if hasattr(engine, 'onion_url'):
-            engine.search_url = engine.onion_url + getattr(engine, 'search_path', '')
-    elif 'onions' in engine.categories:
-        # exclude onion engines if not using tor.
-        return None
+        engine.fetch_supported_languages = (
+            lambda: engine._fetch_supported_languages(
+                get(engine.supported_languages_url, headers=headers))
+        )
 
-    engine.timeout += settings['outgoing'].get('extra_proxy_timeout', 0)
 
-    for category_name in engine.categories:
-        categories.setdefault(category_name, []).append(engine)
+def update_attributes_for_tor(engine):
+    if (settings['outgoing'].get('using_tor_proxy')
+        and hasattr(engine, 'onion_url') ):
+        engine.search_url = engine.onion_url + getattr(engine, 'search_path', '')
+        engine.timeout += settings['outgoing'].get('extra_proxy_timeout', 0)
+
+
+def is_missing_required_attributes(engine):
+    """An attribute is required when its name doesn't start with ``_`` (underline).
+    Required attributes must not be ``None``.
 
+    """
+    missing = False
+    for engine_attr in dir(engine):
+        if not engine_attr.startswith('_') and getattr(engine, engine_attr) is None:
+            logger.error(
+                'Missing engine config attribute: "{0}.{1}"'
+                .format(engine.name, engine_attr))
+            missing = True
+    return missing
+
+
+def is_engine_active(engine):
+    # check if engine is inactive
+    if engine.inactive is True:
+        return False
+
+    # exclude onion engines if not using tor
+    if ('onions' in engine.categories
+        and not settings['outgoing'].get('using_tor_proxy') ):
+        return False
+
+    return True
+
+
+def register_engine(engine):
+    engines[engine.name] = engine
     if engine.shortcut in engine_shortcuts:
         logger.error('Engine config error: ambigious shortcut: {0}'.format(engine.shortcut))
         sys.exit(1)
```

Comment:
Should we remove this sys.exit() also?

Reply:
I'm not sure what should be done, so we can keep sys.exit(1).

But the name has the same issue with the name:
```python
    if engine.name in engines:
        logger.error('Engine config error: ambigious name: {0}'.format(engine.name))
        sys.exit(1)

    if engine.shortcut in engine_shortcuts:
        logger.error('Engine config error: ambigious shortcut: {0}'.format(engine.shortcut))
        sys.exit(1)
```

Reply:
Thanks! / I will rewrite my last commit to check the name also.